### DISCUSSION TITLE
lualine background colors are not being set properly when run from packer

### DISCUSSION THREAD

**Discussion by norseghost** [nvim-lualine/lualine.nvim#627] [Q&A]
I'm not sure if this is a packer issue or a lualine one.

lualine is installed and activated via packer:

```lua
    use { -- statusline
          'nvim-lualine/lualine.nvim',
         -- function wrapper around require(config.statusline)
          config = get_config('statusline')
    }
```

and it works, insofar as i get a lualine; but the section colors are off. seemingly; only the section separators get the updated
 colors.

re-requirin lualine setup fixes it,.

 tested with a minimal rc file (just packer, nord, and lualine) running neovim nightly.

**Comment by fitrh** [comment]
What is `get_config` and what is inside your `config.statusline` ? can you provide a minimal configuration to reproduce the problems ?

**Reply by norseghost** [reply]
> What is `get_config` and what is inside your `config.statusline` ? can you provide a minimal configuration to reproduce the problems ?

I mentioned this in the  code snippet:
a function wrapper to reduce setup function boilerplate in packer. makes no difference if its there or not.
```lua
-- get setup for plugins that need it
local function get_config(name)
    return string.format('require("config/%s")', name)
end
```
minimal rc; with packer autoinstall code withheld for clarity as well as previously discussed wrapper:

```lua
local fn = vim.fn

-- Automatically install packer
local install_path = fn.stdpath('data') .. '/site/pack/packer/start/packer.nvim'
if fn.empty(fn.glob(install_path)) > 0 then
    PACKER_BOOTSTRAP = fn.system({ 'git', 'clone', '--depth', '1', 'https://github.com/wbthomason/packer.nvim', install_path })
    print("Installing packer close and reopen Neovim...")
end
-- Use a protected call so we don't error out on first use
local status_ok, packer = pcall(require, "packer")
if not status_ok then
    return
end

return require('packer').startup(function(use)
    use 'wbthomason/packer.nvim' -- Packer can manage itself

    use { -- color scheme
          'shaunsingh/nord.nvim',
          config = function() require 'nord'.set() end

    }
    use { -- statusline
          'nvim-lualine/lualine.nvim',
          config = function() require 'lualine'.setup() end
    }
    if PACKER_BOOTSTRAP then
        require('packer').sync()
    end

end)

```

....or so  I thought... it is now consistently working as expected (barring a few restarts to me packer happy if starting from complete scratch)

I will dig deeper. may well be packer shenanigans.

**Reply by fitrh** [reply]
You have to wrap your `get_config` call inside an anonymous function

```lua
use {
   'nvim-lualine/lualine.nvim',
   config = function()
      get_config("config.statusline")
   end,
}
```

**Reply by norseghost** [reply]
Huh. Can you point me to an explanation for why that is? The wrapper works as-is for my other plugins; but if that’s causing the issue I might as well remove it and require the setup file in an anymous function directly.

**Reply by fitrh** [reply]
After reading your `get_config` function, the problems is that you are returning a string

The `config` key accepts a string or function, if it is string, it must be a vim command

`require` is lua syntax, so running it as a vim command should be prefixed with `lua`  (e.g. `lua require('module.name')`)

**Reply by fitrh** [reply]
The reason why you need to wrap the function inside an anonymous function is 
because packer will execute the `config` key as a vim command or 
as `config()` (lua function), so when you assign it with 
function call, then the value of `config` is 
the return value of your function

```lua
config = get_config('statusline')
-- it means: config = "require('config/statusline')"
-- it should be: config = "lua require('config/statusline')"
-- which run as vim command
```

If you don't want to wrap it with anonymous function, 
then your function should return a function

```lua
local function get_config(name)
    return function()
        require(("config/%s"):format(name))
    end
end
```

**Reply by norseghost** [reply]
I still don’t get it 

The fuction runs require on the interpolated string parameter — which does get run as lua. 

all other plug-in configs load this way (even lualines; with some color wonkiness the last few days).

**Reply by norseghost** [reply]
Oh ok second reply makes sense 😊
